In [19]:
!pip install --quiet optuna

import optuna
import optuna.integration

     |████████████████████████████████| 308 kB 24.1 MB/s 
     |████████████████████████████████| 81 kB 7.8 MB/s 
     |████████████████████████████████| 209 kB 57.9 MB/s 
     |████████████████████████████████| 78 kB 6.3 MB/s 
     |████████████████████████████████| 112 kB 55.9 MB/s 
     |████████████████████████████████| 49 kB 5.5 MB/s 
     |████████████████████████████████| 146 kB 43.8 MB/s 


In [71]:
import xgboost as xgb
import pandas as pd
import numpy as np
import sklearn
import sklearn.datasets
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.model_selection import train_test_split
from xgboost import XGBClassifier

from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## First  using the evaluation metric "Accuracy"

In [69]:
def objective(trial):
  X = pd.read_csv("/content/drive/My Drive/David_Enoma_PhD_Project/Xsubset.csv",header=None)
  y = pd.read_csv("/content/drive/My Drive/David_Enoma_PhD_Project/hapmap_phenotype_recoded.txt", sep = " ", header = None)
  y.replace([1,2], [0,1], inplace = True)
  X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.33, random_state = 234)
  dtrain = xgb.DMatrix(X_train, label = y_train)
  dtest = xgb.DMatrix(X_test, label = y_test)
  param = {
      "silent": 1,
      "objective": "binary:logistic",
      "eval_metric": "auc",
      # The booster parameter sets the type of learner. Usually this is either a tree 
      # or a linear function. In the case of trees, the model will consist of an ensemble 
      # of trees. For the linear booster, it will be a weighted sum of linear functions.
      "booster": trial.suggest_categorical("booster", ["gbtree", "gblinear", "dart"]),
      "lambda": trial.suggest_loguniform("lambda", 1e-8, 1.0),
      "alpha": trial.suggest_loguniform("alpha", 1e-8, 1.0),
  }

  if param["booster"] == "gbtree" or param["booster"] == "dart":
    param["max_depth"] = trial.suggest_int("max_depth", 1, 9)
    param["eta"]= trial.suggest_loguniform("eta", 1e-8, 1.0)
    param["gamma"]= trial.suggest_loguniform("gamma", 1e-8, 1.0)
    param["grow_policy"] = trial.suggest_categorical("grow_policy", ["depthwise", "lossguide"])

  if param["booster"] == "dart":
    param["sample_type"]=trial.suggest_categorical("sample_type", ["uniform", "weighted"])
    param["normalize_type"]=trial.suggest_categorical("normalize_type", ["tree", "forest"])
    param["rate_drop"] = trial.suggest_loguniform("rate_drop", 1e-8, 1.0)
    param["skip_drop"] = trial.suggest_loguniform("skip_drop", 1e-8, 1.0)


    # Call Back for Pruning

  pruning_callback = optuna.integration.XGBoostPruningCallback(trial, "validation-auc")
  bst = xgb.train(param, dtrain, evals = [(dtest, "validation")], callbacks = [pruning_callback])
  preds = bst.predict(dtest)
  pred_labels = np.rint(preds)
  accuracy = accuracy_score(y_test, pred_labels)
  return accuracy

In [70]:
study = optuna.create_study(direction="maximize")
study.optimize(objective, n_trials = 100, timeout=600)
trial = study.best_trial

print("Accuracy : %f", trial.value)
print("Best HyperParamteters : %f", trial.params)

[I 2022-06-02 17:12:50,838] A new study created in memory with name: no-name-2cf61db9-e1a6-4fe1-a8a4-41fb0e889d71


[0]	validation-auc:0.577922
[1]	validation-auc:0.637987
[2]	validation-auc:0.608766
[3]	validation-auc:0.519481
[4]	validation-auc:0.530844
[5]	validation-auc:0.571429
[6]	validation-auc:0.537338
[7]	validation-auc:0.576299
[8]	validation-auc:0.547078
[9]	validation-auc:0.535714


[I 2022-06-02 17:12:51,024] Trial 0 finished with value: 0.5833333333333334 and parameters: {'booster': 'gbtree', 'lambda': 0.0003447537140857252, 'alpha': 0.0051820331204337705, 'max_depth': 3, 'eta': 1.5086480712144296e-07, 'gamma': 1.1715942103882784e-08, 'grow_policy': 'lossguide'}. Best is trial 0 with value: 0.5833333333333334.


[0]	validation-auc:0.599026
[1]	validation-auc:0.543831
[2]	validation-auc:0.527597
[3]	validation-auc:0.532468
[4]	validation-auc:0.550325
[5]	validation-auc:0.543831
[6]	validation-auc:0.537338
[7]	validation-auc:0.537338
[8]	validation-auc:0.540584
[9]	validation-auc:0.538961


[I 2022-06-02 17:12:51,201] Trial 1 finished with value: 0.5277777777777778 and parameters: {'booster': 'dart', 'lambda': 2.0833130078619587e-08, 'alpha': 5.521644699721967e-07, 'max_depth': 2, 'eta': 0.00037014413803287315, 'gamma': 0.0024492400102343027, 'grow_policy': 'depthwise', 'sample_type': 'weighted', 'normalize_type': 'forest', 'rate_drop': 1.864492287031283e-08, 'skip_drop': 3.4703271164922825e-07}. Best is trial 0 with value: 0.5833333333333334.


[0]	validation-auc:0.548701
[1]	validation-auc:0.561688
[2]	validation-auc:0.594156
[3]	validation-auc:0.633117
[4]	validation-auc:0.642857
[5]	validation-auc:0.652597
[6]	validation-auc:0.662338
[7]	validation-auc:0.659091
[8]	validation-auc:0.652597
[9]	validation-auc:0.655844


[I 2022-06-02 17:12:51,378] Trial 2 finished with value: 0.6111111111111112 and parameters: {'booster': 'gblinear', 'lambda': 0.030475097247617086, 'alpha': 3.753658128510404e-05}. Best is trial 2 with value: 0.6111111111111112.


[0]	validation-auc:0.532468
[1]	validation-auc:0.525974
[2]	validation-auc:0.538961
[3]	validation-auc:0.538961
[4]	validation-auc:0.548701
[5]	validation-auc:0.551948
[6]	validation-auc:0.555195
[7]	validation-auc:0.555195
[8]	validation-auc:0.548701
[9]	validation-auc:0.548701


[I 2022-06-02 17:12:51,550] Trial 3 finished with value: 0.5555555555555556 and parameters: {'booster': 'gblinear', 'lambda': 9.708163639332861e-07, 'alpha': 2.711406647889736e-06}. Best is trial 2 with value: 0.6111111111111112.


[0]	validation-auc:0.472403
[1]	validation-auc:0.464286
[2]	validation-auc:0.363636
[3]	validation-auc:0.363636
[4]	validation-auc:0.396104
[5]	validation-auc:0.418831
[6]	validation-auc:0.392857
[7]	validation-auc:0.37987
[8]	validation-auc:0.405844
[9]	validation-auc:0.38961


[I 2022-06-02 17:12:51,761] Trial 4 finished with value: 0.4166666666666667 and parameters: {'booster': 'gbtree', 'lambda': 1.1606910875535404e-06, 'alpha': 5.956544687881512e-08, 'max_depth': 8, 'eta': 0.12603188490338552, 'gamma': 8.751583443420966e-06, 'grow_policy': 'lossguide'}. Best is trial 2 with value: 0.6111111111111112.
[I 2022-06-02 17:12:51,851] Trial 5 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:51,941] Trial 6 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:52,038] Trial 7 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:52,135] Trial 8 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.592532
[1]	validation-auc:0.548701
[2]	validation-auc:0.605519
[3]	validation-auc:0.545455
[4]	validation-auc:0.545455
[5]	validation-auc:0.542208
[6]	validation-auc:0.545455
[7]	validation-auc:0.540584
[8]	validation-auc:0.547078
[9]	validation-auc:0.542208


[I 2022-06-02 17:12:52,311] Trial 9 finished with value: 0.5833333333333334 and parameters: {'booster': 'dart', 'lambda': 1.313263725069449e-05, 'alpha': 1.2867786141487219e-05, 'max_depth': 3, 'eta': 1.1061901604583867e-06, 'gamma': 1.1833117671371416e-06, 'grow_policy': 'lossguide', 'sample_type': 'uniform', 'normalize_type': 'tree', 'rate_drop': 1.1198022499178081e-06, 'skip_drop': 0.004658182466001781}. Best is trial 2 with value: 0.6111111111111112.
[I 2022-06-02 17:12:52,402] Trial 10 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:52,513] Trial 11 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:52,623] Trial 12 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:52,724] Trial 13 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:52,823] Trial 14 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:52,921] Trial 15 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.568182
[1]	validation-auc:0.571429
[2]	validation-auc:0.603896
[3]	validation-auc:0.613636
[4]	validation-auc:0.636364
[5]	validation-auc:0.636364
[6]	validation-auc:0.646104
[7]	validation-auc:0.655844
[8]	validation-auc:0.646104
[9]	validation-auc:0.642857


[I 2022-06-02 17:12:53,083] Trial 16 finished with value: 0.6111111111111112 and parameters: {'booster': 'gblinear', 'lambda': 0.022074792839411543, 'alpha': 0.00011847069851446123}. Best is trial 2 with value: 0.6111111111111112.


[0]	validation-auc:0.571429
[1]	validation-auc:0.603896
[2]	validation-auc:0.62013
[3]	validation-auc:0.646104
[4]	validation-auc:0.652597
[5]	validation-auc:0.655844
[6]	validation-auc:0.655844
[7]	validation-auc:0.646104
[8]	validation-auc:0.646104
[9]	validation-auc:0.63961


[I 2022-06-02 17:12:53,248] Trial 17 finished with value: 0.6111111111111112 and parameters: {'booster': 'gblinear', 'lambda': 0.05452078776829971, 'alpha': 0.00011011953038025383}. Best is trial 2 with value: 0.6111111111111112.


[0]	validation-auc:0.581169
[1]	validation-auc:0.607143
[2]	validation-auc:0.63961
[3]	validation-auc:0.652597
[4]	validation-auc:0.659091
[5]	validation-auc:0.672078
[6]	validation-auc:0.675325
[7]	validation-auc:0.652597
[8]	validation-auc:0.649351
[9]	validation-auc:0.649351


[I 2022-06-02 17:12:53,421] Trial 18 finished with value: 0.6111111111111112 and parameters: {'booster': 'gblinear', 'lambda': 0.06656521506659605, 'alpha': 1.1695269136146743e-08}. Best is trial 2 with value: 0.6111111111111112.


[0]	validation-auc:0.681818
[1]	validation-auc:0.672078
[2]	validation-auc:0.681818
[3]	validation-auc:0.665584
[4]	validation-auc:0.655844
[5]	validation-auc:0.655844
[6]	validation-auc:0.652597
[7]	validation-auc:0.659091
[8]	validation-auc:0.662338
[9]	validation-auc:0.668831


[I 2022-06-02 17:12:53,596] Trial 19 finished with value: 0.5833333333333334 and parameters: {'booster': 'gblinear', 'lambda': 0.5893888187112625, 'alpha': 1.071604787854792e-08}. Best is trial 2 with value: 0.6111111111111112.


[0]	validation-auc:0.600649
[1]	validation-auc:0.642857
[2]	validation-auc:0.652597
[3]	validation-auc:0.662338
[4]	validation-auc:0.665584
[5]	validation-auc:0.649351
[6]	validation-auc:0.646104
[7]	validation-auc:0.642857
[8]	validation-auc:0.642857
[9]	validation-auc:0.63961


[I 2022-06-02 17:12:53,763] Trial 20 finished with value: 0.6111111111111112 and parameters: {'booster': 'gblinear', 'lambda': 0.12259259435151999, 'alpha': 1.0512663064930394e-05}. Best is trial 2 with value: 0.6111111111111112.


[0]	validation-auc:0.61039
[1]	validation-auc:0.63961
[2]	validation-auc:0.649351
[3]	validation-auc:0.655844
[4]	validation-auc:0.659091
[5]	validation-auc:0.652597
[6]	validation-auc:0.646104
[7]	validation-auc:0.642857
[8]	validation-auc:0.63961
[9]	validation-auc:0.642857


[I 2022-06-02 17:12:53,941] Trial 21 finished with value: 0.6111111111111112 and parameters: {'booster': 'gblinear', 'lambda': 0.1061734455128735, 'alpha': 4.654318884716632e-06}. Best is trial 2 with value: 0.6111111111111112.
[I 2022-06-02 17:12:54,034] Trial 22 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.652597
[1]	validation-auc:0.685065
[2]	validation-auc:0.662338
[3]	validation-auc:0.655844
[4]	validation-auc:0.659091
[5]	validation-auc:0.655844
[6]	validation-auc:0.655844
[7]	validation-auc:0.662338
[8]	validation-auc:0.646104
[9]	validation-auc:0.63961


[I 2022-06-02 17:12:54,213] Trial 23 finished with value: 0.6111111111111112 and parameters: {'booster': 'gblinear', 'lambda': 0.19863152984254126, 'alpha': 1.0083497561034846e-08}. Best is trial 2 with value: 0.6111111111111112.


[0]	validation-auc:0.681818
[1]	validation-auc:0.681818
[2]	validation-auc:0.668831
[3]	validation-auc:0.665584
[4]	validation-auc:0.662338
[5]	validation-auc:0.659091
[6]	validation-auc:0.652597
[7]	validation-auc:0.642857
[8]	validation-auc:0.649351
[9]	validation-auc:0.659091


[I 2022-06-02 17:12:54,386] Trial 24 finished with value: 0.6111111111111112 and parameters: {'booster': 'gblinear', 'lambda': 0.2934855307682667, 'alpha': 2.4130747659781618e-06}. Best is trial 2 with value: 0.6111111111111112.
[I 2022-06-02 17:12:54,489] Trial 25 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:54,609] Trial 26 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:54,711] Trial 27 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.678571
[1]	validation-auc:0.668831
[2]	validation-auc:0.665584
[3]	validation-auc:0.655844
[4]	validation-auc:0.655844
[5]	validation-auc:0.655844
[6]	validation-auc:0.649351
[7]	validation-auc:0.646104
[8]	validation-auc:0.652597
[9]	validation-auc:0.652597


[I 2022-06-02 17:12:54,893] Trial 28 finished with value: 0.6111111111111112 and parameters: {'booster': 'gblinear', 'lambda': 0.33121550493555046, 'alpha': 1.2766415173234674e-06}. Best is trial 2 with value: 0.6111111111111112.


[0]	validation-auc:0.665584
[1]	validation-auc:0.681818
[2]	validation-auc:0.678571
[3]	validation-auc:0.672078
[4]	validation-auc:0.662338
[5]	validation-auc:0.655844
[6]	validation-auc:0.659091
[7]	validation-auc:0.662338
[8]	validation-auc:0.659091
[9]	validation-auc:0.665584


[I 2022-06-02 17:12:55,074] Trial 29 finished with value: 0.5833333333333334 and parameters: {'booster': 'gblinear', 'lambda': 0.4603704176629928, 'alpha': 1.3247231496751098e-06}. Best is trial 2 with value: 0.6111111111111112.
[I 2022-06-02 17:12:55,171] Trial 30 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.665584
[1]	validation-auc:0.675325
[2]	validation-auc:0.675325
[3]	validation-auc:0.665584
[4]	validation-auc:0.652597
[5]	validation-auc:0.649351
[6]	validation-auc:0.649351
[7]	validation-auc:0.652597
[8]	validation-auc:0.652597
[9]	validation-auc:0.652597


[I 2022-06-02 17:12:55,355] Trial 31 finished with value: 0.5833333333333334 and parameters: {'booster': 'gblinear', 'lambda': 0.4310783912865813, 'alpha': 2.391673678561165e-05}. Best is trial 2 with value: 0.6111111111111112.


[0]	validation-auc:0.659091
[1]	validation-auc:0.672078
[2]	validation-auc:0.662338
[3]	validation-auc:0.662338
[4]	validation-auc:0.659091
[5]	validation-auc:0.659091
[6]	validation-auc:0.652597
[7]	validation-auc:0.649351
[8]	validation-auc:0.646104
[9]	validation-auc:0.646104


[I 2022-06-02 17:12:55,536] Trial 32 finished with value: 0.5833333333333334 and parameters: {'booster': 'gblinear', 'lambda': 0.2611728052489749, 'alpha': 2.7509418015174816e-08}. Best is trial 2 with value: 0.6111111111111112.
[I 2022-06-02 17:12:55,639] Trial 33 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:55,735] Trial 34 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:55,828] Trial 35 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:55,930] Trial 36 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:56,025] Trial 37 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:56,143] Trial 38 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.681818
[1]	validation-auc:0.681818
[2]	validation-auc:0.685065
[3]	validation-auc:0.672078
[4]	validation-auc:0.675325
[5]	validation-auc:0.675325
[6]	validation-auc:0.672078
[7]	validation-auc:0.672078
[8]	validation-auc:0.672078
[9]	validation-auc:0.688312


[I 2022-06-02 17:12:56,319] Trial 39 finished with value: 0.6111111111111112 and parameters: {'booster': 'gblinear', 'lambda': 0.9331465381844356, 'alpha': 3.3893749580978375e-06}. Best is trial 2 with value: 0.6111111111111112.


[0]	validation-auc:0.672078
[1]	validation-auc:0.678571
[2]	validation-auc:0.678571
[3]	validation-auc:0.668831
[4]	validation-auc:0.665584
[5]	validation-auc:0.668831
[6]	validation-auc:0.672078
[7]	validation-auc:0.665584
[8]	validation-auc:0.665584
[9]	validation-auc:0.672078


[I 2022-06-02 17:12:56,504] Trial 40 finished with value: 0.5833333333333334 and parameters: {'booster': 'gblinear', 'lambda': 0.7512314344715854, 'alpha': 8.447544422956306e-06}. Best is trial 2 with value: 0.6111111111111112.


[0]	validation-auc:0.613636
[1]	validation-auc:0.652597
[2]	validation-auc:0.665584
[3]	validation-auc:0.665584
[4]	validation-auc:0.655844
[5]	validation-auc:0.63961
[6]	validation-auc:0.642857
[7]	validation-auc:0.646104
[8]	validation-auc:0.63961
[9]	validation-auc:0.636364


[I 2022-06-02 17:12:56,679] Trial 41 finished with value: 0.6111111111111112 and parameters: {'booster': 'gblinear', 'lambda': 0.1465681789619733, 'alpha': 1.9858334098599276e-05}. Best is trial 2 with value: 0.6111111111111112.


[0]	validation-auc:0.652597
[1]	validation-auc:0.675325
[2]	validation-auc:0.672078
[3]	validation-auc:0.675325
[4]	validation-auc:0.665584
[5]	validation-auc:0.662338
[6]	validation-auc:0.652597
[7]	validation-auc:0.649351
[8]	validation-auc:0.649351
[9]	validation-auc:0.646104


[I 2022-06-02 17:12:56,863] Trial 42 finished with value: 0.6111111111111112 and parameters: {'booster': 'gblinear', 'lambda': 0.2567842716919653, 'alpha': 3.054036352992012e-06}. Best is trial 2 with value: 0.6111111111111112.
[I 2022-06-02 17:12:56,959] Trial 43 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.613636


[I 2022-06-02 17:12:57,063] Trial 44 pruned. Trial was pruned at iteration 1.
[I 2022-06-02 17:12:57,160] Trial 45 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:57,267] Trial 46 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:57,362] Trial 47 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:57,467] Trial 48 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.646104
[1]	validation-auc:0.668831
[2]	validation-auc:0.672078
[3]	validation-auc:0.665584
[4]	validation-auc:0.655844
[5]	validation-auc:0.665584
[6]	validation-auc:0.659091
[7]	validation-auc:0.646104
[8]	validation-auc:0.652597
[9]	validation-auc:0.646104


[I 2022-06-02 17:12:57,649] Trial 49 finished with value: 0.6111111111111112 and parameters: {'booster': 'gblinear', 'lambda': 0.16865589219627294, 'alpha': 1.6883506586933825e-07}. Best is trial 2 with value: 0.6111111111111112.


[0]	validation-auc:0.662338
[1]	validation-auc:0.672078
[2]	validation-auc:0.662338
[3]	validation-auc:0.655844
[4]	validation-auc:0.646104
[5]	validation-auc:0.649351
[6]	validation-auc:0.646104
[7]	validation-auc:0.642857
[8]	validation-auc:0.649351
[9]	validation-auc:0.649351


[I 2022-06-02 17:12:57,828] Trial 50 finished with value: 0.5833333333333334 and parameters: {'booster': 'gblinear', 'lambda': 0.37289073887443636, 'alpha': 2.0324268293877632e-06}. Best is trial 2 with value: 0.6111111111111112.
[I 2022-06-02 17:12:57,921] Trial 51 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.685065
[1]	validation-auc:0.675325
[2]	validation-auc:0.675325
[3]	validation-auc:0.665584
[4]	validation-auc:0.659091
[5]	validation-auc:0.665584
[6]	validation-auc:0.665584
[7]	validation-auc:0.665584
[8]	validation-auc:0.665584
[9]	validation-auc:0.672078


[I 2022-06-02 17:12:58,101] Trial 52 finished with value: 0.5833333333333334 and parameters: {'booster': 'gblinear', 'lambda': 0.660157118881854, 'alpha': 1.7988721347881815e-05}. Best is trial 2 with value: 0.6111111111111112.
[I 2022-06-02 17:12:58,214] Trial 53 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:58,327] Trial 54 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:58,422] Trial 55 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:58,514] Trial 56 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.659091
[1]	validation-auc:0.672078
[2]	validation-auc:0.662338
[3]	validation-auc:0.675325
[4]	validation-auc:0.655844
[5]	validation-auc:0.655844
[6]	validation-auc:0.649351
[7]	validation-auc:0.642857
[8]	validation-auc:0.646104
[9]	validation-auc:0.63961


[I 2022-06-02 17:12:58,696] Trial 57 finished with value: 0.6111111111111112 and parameters: {'booster': 'gblinear', 'lambda': 0.25143263091776985, 'alpha': 0.00013446709709358945}. Best is trial 2 with value: 0.6111111111111112.


[0]	validation-auc:0.675325
[1]	validation-auc:0.675325
[2]	validation-auc:0.688312
[3]	validation-auc:0.681818
[4]	validation-auc:0.675325
[5]	validation-auc:0.675325
[6]	validation-auc:0.672078
[7]	validation-auc:0.688312
[8]	validation-auc:0.691558
[9]	validation-auc:0.691558


[I 2022-06-02 17:12:58,867] Trial 58 finished with value: 0.5833333333333334 and parameters: {'booster': 'gblinear', 'lambda': 0.9034103748244214, 'alpha': 3.6504416779762378e-06}. Best is trial 2 with value: 0.6111111111111112.
[I 2022-06-02 17:12:58,975] Trial 59 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:59,073] Trial 60 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:59,164] Trial 61 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:59,253] Trial 62 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:59,349] Trial 63 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.681818
[1]	validation-auc:0.678571
[2]	validation-auc:0.662338
[3]	validation-auc:0.662338
[4]	validation-auc:0.655844
[5]	validation-auc:0.655844
[6]	validation-auc:0.649351
[7]	validation-auc:0.649351
[8]	validation-auc:0.652597
[9]	validation-auc:0.652597


[I 2022-06-02 17:12:59,525] Trial 64 finished with value: 0.5833333333333334 and parameters: {'booster': 'gblinear', 'lambda': 0.3575168599466011, 'alpha': 4.8088685005662895e-08}. Best is trial 2 with value: 0.6111111111111112.
[I 2022-06-02 17:12:59,629] Trial 65 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:59,731] Trial 66 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:59,827] Trial 67 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:12:59,934] Trial 68 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:13:00,030] Trial 69 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.678571
[1]	validation-auc:0.665584
[2]	validation-auc:0.672078
[3]	validation-auc:0.672078
[4]	validation-auc:0.678571
[5]	validation-auc:0.688312
[6]	validation-auc:0.688312
[7]	validation-auc:0.694805
[8]	validation-auc:0.698052
[9]	validation-auc:0.698052


[I 2022-06-02 17:13:00,213] Trial 70 finished with value: 0.5833333333333334 and parameters: {'booster': 'gblinear', 'lambda': 0.9072948410507382, 'alpha': 2.3992907664011372e-06}. Best is trial 2 with value: 0.6111111111111112.
[I 2022-06-02 17:13:00,304] Trial 71 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:13:00,404] Trial 72 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:13:00,500] Trial 73 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:13:00,600] Trial 74 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:13:00,734] Trial 75 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.691558
[1]	validation-auc:0.665584
[2]	validation-auc:0.678571
[3]	validation-auc:0.659091
[4]	validation-auc:0.655844
[5]	validation-auc:0.655844
[6]	validation-auc:0.655844
[7]	validation-auc:0.649351
[8]	validation-auc:0.655844
[9]	validation-auc:0.665584


[I 2022-06-02 17:13:00,920] Trial 76 finished with value: 0.5833333333333334 and parameters: {'booster': 'gblinear', 'lambda': 0.5484238763014397, 'alpha': 3.4534215773212926e-07}. Best is trial 2 with value: 0.6111111111111112.
[I 2022-06-02 17:13:01,012] Trial 77 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:13:01,130] Trial 78 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:13:01,230] Trial 79 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.668831
[1]	validation-auc:0.685065
[2]	validation-auc:0.665584
[3]	validation-auc:0.649351
[4]	validation-auc:0.655844
[5]	validation-auc:0.649351
[6]	validation-auc:0.649351
[7]	validation-auc:0.649351
[8]	validation-auc:0.655844
[9]	validation-auc:0.652597


[I 2022-06-02 17:13:01,413] Trial 80 finished with value: 0.6111111111111112 and parameters: {'booster': 'gblinear', 'lambda': 0.34619035635638506, 'alpha': 8.356804397645215e-07}. Best is trial 2 with value: 0.6111111111111112.
[I 2022-06-02 17:13:01,512] Trial 81 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:13:01,610] Trial 82 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:13:01,714] Trial 83 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.672078
[1]	validation-auc:0.701299
[2]	validation-auc:0.688312
[3]	validation-auc:0.688312
[4]	validation-auc:0.672078
[5]	validation-auc:0.675325
[6]	validation-auc:0.668831
[7]	validation-auc:0.655844
[8]	validation-auc:0.649351
[9]	validation-auc:0.649351


[I 2022-06-02 17:13:01,898] Trial 84 finished with value: 0.6111111111111112 and parameters: {'booster': 'gblinear', 'lambda': 0.16620263128538051, 'alpha': 1.7596444305966254e-08}. Best is trial 2 with value: 0.6111111111111112.


[0]	validation-auc:0.668831
[1]	validation-auc:0.672078
[2]	validation-auc:0.659091
[3]	validation-auc:0.649351
[4]	validation-auc:0.659091
[5]	validation-auc:0.655844
[6]	validation-auc:0.649351
[7]	validation-auc:0.649351
[8]	validation-auc:0.652597
[9]	validation-auc:0.649351


[I 2022-06-02 17:13:02,068] Trial 85 finished with value: 0.6111111111111112 and parameters: {'booster': 'gblinear', 'lambda': 0.2878274934072311, 'alpha': 6.703913351037345e-07}. Best is trial 2 with value: 0.6111111111111112.
[I 2022-06-02 17:13:02,163] Trial 86 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.688312
[1]	validation-auc:0.685065
[2]	validation-auc:0.678571
[3]	validation-auc:0.668831
[4]	validation-auc:0.665584
[5]	validation-auc:0.668831
[6]	validation-auc:0.665584
[7]	validation-auc:0.662338
[8]	validation-auc:0.672078
[9]	validation-auc:0.665584


[I 2022-06-02 17:13:02,345] Trial 87 finished with value: 0.5833333333333334 and parameters: {'booster': 'gblinear', 'lambda': 0.5842922472358862, 'alpha': 6.269260721479115e-08}. Best is trial 2 with value: 0.6111111111111112.
[I 2022-06-02 17:13:02,437] Trial 88 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:13:02,540] Trial 89 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:13:02,673] Trial 90 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:13:02,769] Trial 91 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:13:02,864] Trial 92 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.688312
[1]	validation-auc:0.678571
[2]	validation-auc:0.672078
[3]	validation-auc:0.672078
[4]	validation-auc:0.675325
[5]	validation-auc:0.678571
[6]	validation-auc:0.678571
[7]	validation-auc:0.672078
[8]	validation-auc:0.675325
[9]	validation-auc:0.685065


[I 2022-06-02 17:13:03,050] Trial 93 finished with value: 0.5833333333333334 and parameters: {'booster': 'gblinear', 'lambda': 0.9137835636225653, 'alpha': 3.929348201199006e-06}. Best is trial 2 with value: 0.6111111111111112.
[I 2022-06-02 17:13:03,144] Trial 94 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.668831
[1]	validation-auc:0.675325
[2]	validation-auc:0.681818
[3]	validation-auc:0.665584
[4]	validation-auc:0.652597
[5]	validation-auc:0.655844
[6]	validation-auc:0.659091
[7]	validation-auc:0.662338
[8]	validation-auc:0.662338
[9]	validation-auc:0.662338


[I 2022-06-02 17:13:03,320] Trial 95 finished with value: 0.5833333333333334 and parameters: {'booster': 'gblinear', 'lambda': 0.5078823815432375, 'alpha': 1.4413684163468104e-08}. Best is trial 2 with value: 0.6111111111111112.
[I 2022-06-02 17:13:03,415] Trial 96 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:13:03,521] Trial 97 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:13:03,622] Trial 98 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:13:03,730] Trial 99 pruned. Trial was pruned at iteration 0.


Accuracy : %f 0.6111111111111112
Best HyperParamteters : %f {'booster': 'gblinear', 'lambda': 0.030475097247617086, 'alpha': 3.753658128510404e-05}


### Accuracy 
0.6111111111111112

### Best HyperParamteters 
'booster': 'gblinear', 

'lambda': 0.030475097247617086, 

'alpha': 3.753658128510404e-05

## For a problem as sensitive as Cancer, Evaluation metrics such as "Recall" or "F1 Score" are more important than "Accuracy" because the focus is on ensuring the highest number of people who have cancer are predicted to have cancer and so I will be using "Recall" as my evaluation metric here

In [74]:
def objective(trial):
  X = pd.read_csv("/content/drive/My Drive/David_Enoma_PhD_Project/Xsubset.csv",header=None)
  y = pd.read_csv("/content/drive/My Drive/David_Enoma_PhD_Project/hapmap_phenotype_recoded.txt", sep = " ", header = None)
  y.replace([1,2], [0,1], inplace = True)
  X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.33, random_state = 234)
  dtrain = xgb.DMatrix(X_train, label = y_train)
  dtest = xgb.DMatrix(X_test, label = y_test)
  param = {
      "verbosity": 0,
      # logistic regression for binary classification, output probability
      "objective": "binary:logistic",
      # use exact for small dataset.
      "tree_method": "auto",
      "eval_metric": "auc",
      # The booster parameter sets the type of learner. Usually this is either a tree 
      # or a linear function. In the case of trees, the model will consist of an ensemble 
      # of trees. For the linear booster, it will be a weighted sum of linear functions.
      # defines booster, gblinear for linear functions.
      "booster": trial.suggest_categorical("booster", ["gbtree", "gblinear", "dart"]),
      # L2 regularization weight.
      "lambda": trial.suggest_float("lambda", 1e-8, 1.0, log = True), 
      # L1 regularization weight.
      "alpha": trial.suggest_float("alpha", 1e-8, 1.0, log = True),
      # sampling ratio for training data.
      "subsample": trial.suggest_float("subsample", 0.2, 1.0),
      # sampling according to each tree.
      "colsample_bytree": trial.suggest_float("colsample_bytree", 0.2, 1.0),
  }

  if param["booster"] == "gbtree" or param["booster"] == "dart":
    # maximum depth of the tree, signifies complexity of the tree.
    param["max_depth"] = trial.suggest_int("max_depth", 1, 7, step=2)
    # minimum child weight, larger the term more conservative the tree.
    param["min_child_weight"] = trial.suggest_int("min_child_weight", 2, 10)
    # "eta" shrinks the feature weights to make the boosting process more conservative.
    param["eta"]= trial.suggest_float("eta", 1e-8, 1.0, log = True)
    # Minimum loss reduction required to make a further partition on a leaf node of the tree. 
    # The larger gamma is, the more conservative the algorithm will be.
    param["gamma"]= trial.suggest_float("gamma", 1e-8, 1.0, log = True)
    #  Controls the way new nodes are added to the tree.
    param["grow_policy"] = trial.suggest_categorical("grow_policy", ["depthwise", "lossguide"])
    
  if param["booster"] == "dart":
    # Type of sampling algorithm.
    param["sample_type"]=trial.suggest_categorical("sample_type", ["uniform", "weighted"])
    # Type of normalization algorithm.
    param["normalize_type"]=trial.suggest_categorical("normalize_type", ["tree", "forest"])
    # Dropout rate (a fraction of previous trees to drop during the dropout).
    param["rate_drop"] = trial.suggest_float("rate_drop", 1e-8, 1.0, log = True)
    # Probability of skipping the dropout procedure during a boosting iteration.
    param["skip_drop"] = trial.suggest_float("skip_drop", 1e-8, 1.0, log = True)


  # Call Back for Pruning
  # Pruning is used to terminate unpromising trials early, so that computing time can be used for trials that show more potential.
  pruning_callback = optuna.integration.XGBoostPruningCallback(trial, "validation-auc")
  bst = xgb.train(param, dtrain, evals = [(dtest, "validation")], callbacks = [pruning_callback])
  preds = bst.predict(dtest)
  # np.rint means return integer. It rounds up elements to the nearest integer. 
  # This is done so we can use accuracy_score because "binary:logistic" outputs probability not discrete numbers [0,1] but the probability of each sample being 1
  pred_labels = np.rint(preds)
  Recall = recall_score(y_test, pred_labels)
  return Recall

In [73]:
study = optuna.create_study(direction="maximize")
study.optimize(objective, n_trials = 100, timeout=600)
trial = study.best_trial

print("Recall : %f", trial.value)
print("Best HyperParamteters : %f", trial.params)

[I 2022-06-02 17:38:47,803] A new study created in memory with name: no-name-51a4781b-0934-418f-a61f-a177cffb35d4


[0]	validation-auc:0.626623
[1]	validation-auc:0.63961
[2]	validation-auc:0.62987
[3]	validation-auc:0.603896
[4]	validation-auc:0.594156
[5]	validation-auc:0.600649
[6]	validation-auc:0.587662
[7]	validation-auc:0.581169
[8]	validation-auc:0.584416
[9]	validation-auc:0.577922


[I 2022-06-02 17:38:48,156] Trial 0 finished with value: 0.7857142857142857 and parameters: {'booster': 'gblinear', 'lambda': 5.491115673572802e-08, 'alpha': 0.002810245821911446, 'subsample': 0.587167981915971, 'colsample_bytree': 0.37519211627122084}. Best is trial 0 with value: 0.7857142857142857.


[0]	validation-auc:0.548701
[1]	validation-auc:0.571429
[2]	validation-auc:0.623377
[3]	validation-auc:0.626623
[4]	validation-auc:0.631494
[5]	validation-auc:0.623377
[6]	validation-auc:0.616883
[7]	validation-auc:0.61526
[8]	validation-auc:0.618506
[9]	validation-auc:0.568182


[I 2022-06-02 17:38:48,503] Trial 1 finished with value: 0.7142857142857143 and parameters: {'booster': 'dart', 'lambda': 0.00027626960218040563, 'alpha': 0.5867635045050236, 'subsample': 0.6608825479286826, 'colsample_bytree': 0.38627499741087096, 'max_depth': 1, 'min_child_weight': 5, 'eta': 1.4474382771489371e-06, 'gamma': 4.5679746479102705e-07, 'grow_policy': 'lossguide', 'sample_type': 'weighted', 'normalize_type': 'forest', 'rate_drop': 0.008106881402335753, 'skip_drop': 0.013935454738290267}. Best is trial 0 with value: 0.7857142857142857.


[0]	validation-auc:0.5
[1]	validation-auc:0.5
[2]	validation-auc:0.5
[3]	validation-auc:0.5
[4]	validation-auc:0.5
[5]	validation-auc:0.5
[6]	validation-auc:0.5
[7]	validation-auc:0.5
[8]	validation-auc:0.5
[9]	validation-auc:0.5


[I 2022-06-02 17:38:48,893] Trial 2 finished with value: 1.0 and parameters: {'booster': 'dart', 'lambda': 8.669790267915281e-07, 'alpha': 1.1536382592082632e-08, 'subsample': 0.28469669085250404, 'colsample_bytree': 0.5566590888822147, 'max_depth': 7, 'min_child_weight': 10, 'eta': 0.003933096882701822, 'gamma': 0.0018015367008443472, 'grow_policy': 'lossguide', 'sample_type': 'weighted', 'normalize_type': 'forest', 'rate_drop': 8.829825957910605e-06, 'skip_drop': 5.189332052983798e-07}. Best is trial 2 with value: 1.0.


[0]	validation-auc:0.545455
[1]	validation-auc:0.548701
[2]	validation-auc:0.532468
[3]	validation-auc:0.542208
[4]	validation-auc:0.542208
[5]	validation-auc:0.555195
[6]	validation-auc:0.548701
[7]	validation-auc:0.574675
[8]	validation-auc:0.564935
[9]	validation-auc:0.568182


[I 2022-06-02 17:38:49,323] Trial 3 finished with value: 0.7142857142857143 and parameters: {'booster': 'gblinear', 'lambda': 2.0014179227598036e-07, 'alpha': 0.0019411170353898178, 'subsample': 0.8554244416963153, 'colsample_bytree': 0.32293255318158204}. Best is trial 2 with value: 1.0.


[0]	validation-auc:0.483766
[1]	validation-auc:0.483766
[2]	validation-auc:0.483766
[3]	validation-auc:0.483766
[4]	validation-auc:0.483766
[5]	validation-auc:0.483766
[6]	validation-auc:0.483766
[7]	validation-auc:0.483766
[8]	validation-auc:0.483766
[9]	validation-auc:0.483766


[I 2022-06-02 17:38:49,813] Trial 4 finished with value: 1.0 and parameters: {'booster': 'gblinear', 'lambda': 5.418231021152819e-07, 'alpha': 0.125013443937533, 'subsample': 0.5233839112296172, 'colsample_bytree': 0.37790130961603396}. Best is trial 2 with value: 1.0.
[I 2022-06-02 17:38:49,956] Trial 5 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:50,046] Trial 6 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:50,144] Trial 7 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:50,235] Trial 8 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:50,339] Trial 9 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:50,435] Trial 10 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:50,534] Trial 11 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:50,645] Trial 12 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:50,739] Trial 13 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:50,837] Trial

[0]	validation-auc:0.613636
[1]	validation-auc:0.600649
[2]	validation-auc:0.597403
[3]	validation-auc:0.613636
[4]	validation-auc:0.61039
[5]	validation-auc:0.62013
[6]	validation-auc:0.62013
[7]	validation-auc:0.61039
[8]	validation-auc:0.607143
[9]	validation-auc:0.607143


[I 2022-06-02 17:38:51,125] Trial 16 finished with value: 0.7857142857142857 and parameters: {'booster': 'gblinear', 'lambda': 0.0025623013628833908, 'alpha': 0.038304873654782315, 'subsample': 0.310075306858276, 'colsample_bytree': 0.22319930018773704}. Best is trial 2 with value: 1.0.
[I 2022-06-02 17:38:51,229] Trial 17 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:51,350] Trial 18 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:51,453] Trial 19 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:51,561] Trial 20 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:51,655] Trial 21 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:51,755] Trial 22 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:51,854] Trial 23 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.63961
[1]	validation-auc:0.626623
[2]	validation-auc:0.594156
[3]	validation-auc:0.581169
[4]	validation-auc:0.581169
[5]	validation-auc:0.574675
[6]	validation-auc:0.568182
[7]	validation-auc:0.571429
[8]	validation-auc:0.571429
[9]	validation-auc:0.574675


[I 2022-06-02 17:38:52,036] Trial 24 finished with value: 0.7857142857142857 and parameters: {'booster': 'gblinear', 'lambda': 7.053076458613763e-08, 'alpha': 0.019269924916397646, 'subsample': 0.7566444514059856, 'colsample_bytree': 0.5169792034325816}. Best is trial 2 with value: 1.0.
[I 2022-06-02 17:38:52,137] Trial 25 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.548701


[I 2022-06-02 17:38:52,243] Trial 26 pruned. Trial was pruned at iteration 1.


[0]	validation-auc:0.564935
[1]	validation-auc:0.623377
[2]	validation-auc:0.650974
[3]	validation-auc:0.628247
[4]	validation-auc:0.631494
[5]	validation-auc:0.646104
[6]	validation-auc:0.63961
[7]	validation-auc:0.616883
[8]	validation-auc:0.633117
[9]	validation-auc:0.600649


[I 2022-06-02 17:38:52,443] Trial 27 finished with value: 0.7142857142857143 and parameters: {'booster': 'dart', 'lambda': 1.7431706756521551e-06, 'alpha': 0.13164652696365411, 'subsample': 0.8239169186101118, 'colsample_bytree': 0.27038286214224694, 'max_depth': 5, 'min_child_weight': 4, 'eta': 0.0023745471866226694, 'gamma': 0.058363249579821475, 'grow_policy': 'depthwise', 'sample_type': 'weighted', 'normalize_type': 'forest', 'rate_drop': 5.746923986065874e-07, 'skip_drop': 0.0011097963731300856}. Best is trial 2 with value: 1.0.


[0]	validation-auc:0.691558
[1]	validation-auc:0.678571
[2]	validation-auc:0.668831
[3]	validation-auc:0.672078
[4]	validation-auc:0.662338
[5]	validation-auc:0.659091
[6]	validation-auc:0.665584
[7]	validation-auc:0.668831
[8]	validation-auc:0.668831
[9]	validation-auc:0.668831


[I 2022-06-02 17:38:52,620] Trial 28 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.5099902582608936, 'alpha': 7.508381173322264e-08, 'subsample': 0.4778861738772048, 'colsample_bytree': 0.49700610305245213}. Best is trial 2 with value: 1.0.
[I 2022-06-02 17:38:52,731] Trial 29 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:52,829] Trial 30 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:52,925] Trial 31 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:53,022] Trial 32 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:53,118] Trial 33 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:53,217] Trial 34 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.675325
[1]	validation-auc:0.675325
[2]	validation-auc:0.665584
[3]	validation-auc:0.655844
[4]	validation-auc:0.665584
[5]	validation-auc:0.668831
[6]	validation-auc:0.668831
[7]	validation-auc:0.675325
[8]	validation-auc:0.672078
[9]	validation-auc:0.675325


[I 2022-06-02 17:38:53,399] Trial 35 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.7774978596810086, 'alpha': 0.0015097756939350465, 'subsample': 0.29824154966884187, 'colsample_bytree': 0.40968138408797145}. Best is trial 2 with value: 1.0.


[0]	validation-auc:0.681818
[1]	validation-auc:0.678571
[2]	validation-auc:0.675325
[3]	validation-auc:0.662338
[4]	validation-auc:0.655844
[5]	validation-auc:0.655844
[6]	validation-auc:0.659091
[7]	validation-auc:0.662338
[8]	validation-auc:0.665584
[9]	validation-auc:0.668831


[I 2022-06-02 17:38:53,578] Trial 36 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.5657583824188351, 'alpha': 3.5784860943660413e-05, 'subsample': 0.21816358595044083, 'colsample_bytree': 0.5801249230615527}. Best is trial 2 with value: 1.0.
[I 2022-06-02 17:38:53,693] Trial 37 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.675325
[1]	validation-auc:0.668831
[2]	validation-auc:0.675325
[3]	validation-auc:0.662338
[4]	validation-auc:0.662338
[5]	validation-auc:0.665584
[6]	validation-auc:0.659091
[7]	validation-auc:0.675325
[8]	validation-auc:0.678571
[9]	validation-auc:0.694805


[I 2022-06-02 17:38:53,874] Trial 38 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.8678339767095239, 'alpha': 0.0016880881614473096, 'subsample': 0.20861789522989502, 'colsample_bytree': 0.6172252728841277}. Best is trial 2 with value: 1.0.
[I 2022-06-02 17:38:53,974] Trial 39 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:54,076] Trial 40 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.63961
[1]	validation-auc:0.636364
[2]	validation-auc:0.633117
[3]	validation-auc:0.642857
[4]	validation-auc:0.646104
[5]	validation-auc:0.642857
[6]	validation-auc:0.642857
[7]	validation-auc:0.646104
[8]	validation-auc:0.646104
[9]	validation-auc:0.649351


[I 2022-06-02 17:38:54,251] Trial 41 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.6896071363151052, 'alpha': 0.01424316019229061, 'subsample': 0.2852170123704754, 'colsample_bytree': 0.656051255700898}. Best is trial 2 with value: 1.0.


[0]	validation-auc:0.642857
[1]	validation-auc:0.662338
[2]	validation-auc:0.646104
[3]	validation-auc:0.646104
[4]	validation-auc:0.636364
[5]	validation-auc:0.636364
[6]	validation-auc:0.633117
[7]	validation-auc:0.623377
[8]	validation-auc:0.623377
[9]	validation-auc:0.623377


[I 2022-06-02 17:38:54,431] Trial 42 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.21381298692096878, 'alpha': 0.001944750969552748, 'subsample': 0.2062076176478566, 'colsample_bytree': 0.6503185002107346}. Best is trial 2 with value: 1.0.
[I 2022-06-02 17:38:54,530] Trial 43 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.659091
[1]	validation-auc:0.646104
[2]	validation-auc:0.642857
[3]	validation-auc:0.659091
[4]	validation-auc:0.659091
[5]	validation-auc:0.659091
[6]	validation-auc:0.662338
[7]	validation-auc:0.672078
[8]	validation-auc:0.672078
[9]	validation-auc:0.681818


[I 2022-06-02 17:38:54,704] Trial 44 finished with value: 0.9285714285714286 and parameters: {'booster': 'gblinear', 'lambda': 0.9355151107440404, 'alpha': 0.018511052695539945, 'subsample': 0.2763104234993489, 'colsample_bytree': 0.6097413325062915}. Best is trial 2 with value: 1.0.


[0]	validation-auc:0.63961
[1]	validation-auc:0.633117
[2]	validation-auc:0.616883
[3]	validation-auc:0.62013
[4]	validation-auc:0.623377
[5]	validation-auc:0.613636
[6]	validation-auc:0.616883
[7]	validation-auc:0.616883
[8]	validation-auc:0.616883
[9]	validation-auc:0.616883


[I 2022-06-02 17:38:54,887] Trial 45 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.22143750340665014, 'alpha': 0.020601356878927984, 'subsample': 0.2763873219460379, 'colsample_bytree': 0.6616958446363022}. Best is trial 2 with value: 1.0.
[I 2022-06-02 17:38:54,991] Trial 46 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:55,088] Trial 47 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:55,182] Trial 48 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:55,284] Trial 49 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:55,391] Trial 50 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:55,495] Trial 51 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.652597
[1]	validation-auc:0.662338
[2]	validation-auc:0.646104
[3]	validation-auc:0.652597
[4]	validation-auc:0.646104
[5]	validation-auc:0.636364
[6]	validation-auc:0.63961
[7]	validation-auc:0.626623
[8]	validation-auc:0.62013
[9]	validation-auc:0.623377


[I 2022-06-02 17:38:55,679] Trial 52 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.3362658241724193, 'alpha': 0.002359103409629776, 'subsample': 0.30517456732524345, 'colsample_bytree': 0.4566619635666215}. Best is trial 2 with value: 1.0.
[I 2022-06-02 17:38:55,780] Trial 53 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:55,880] Trial 54 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:55,975] Trial 55 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:56,073] Trial 56 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:56,185] Trial 57 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:56,278] Trial 58 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:56,375] Trial 59 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:56,488] Trial 60 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.681818
[1]	validation-auc:0.665584
[2]	validation-auc:0.668831
[3]	validation-auc:0.665584
[4]	validation-auc:0.675325
[5]	validation-auc:0.681818
[6]	validation-auc:0.678571
[7]	validation-auc:0.678571
[8]	validation-auc:0.681818
[9]	validation-auc:0.691558


[I 2022-06-02 17:38:56,673] Trial 61 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.9471728537334849, 'alpha': 9.403727624999125e-05, 'subsample': 0.22948790261960694, 'colsample_bytree': 0.571581456638795}. Best is trial 2 with value: 1.0.


[0]	validation-auc:0.642857
[1]	validation-auc:0.665584
[2]	validation-auc:0.649351
[3]	validation-auc:0.646104
[4]	validation-auc:0.649351
[5]	validation-auc:0.633117
[6]	validation-auc:0.62987
[7]	validation-auc:0.62987
[8]	validation-auc:0.623377
[9]	validation-auc:0.62013


[I 2022-06-02 17:38:56,853] Trial 62 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.24922499933221642, 'alpha': 0.002927555749250313, 'subsample': 0.3289336759846834, 'colsample_bytree': 0.5878102503760372}. Best is trial 2 with value: 1.0.
[I 2022-06-02 17:38:56,944] Trial 63 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.688312
[1]	validation-auc:0.672078
[2]	validation-auc:0.685065
[3]	validation-auc:0.681818
[4]	validation-auc:0.675325
[5]	validation-auc:0.681818
[6]	validation-auc:0.678571
[7]	validation-auc:0.681818
[8]	validation-auc:0.688312
[9]	validation-auc:0.694805


[I 2022-06-02 17:38:57,122] Trial 64 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.9491055894582362, 'alpha': 3.211185756794371e-08, 'subsample': 0.29901641587613426, 'colsample_bytree': 0.5859596818527173}. Best is trial 2 with value: 1.0.


[0]	validation-auc:0.675325
[1]	validation-auc:0.665584
[2]	validation-auc:0.678571
[3]	validation-auc:0.675325
[4]	validation-auc:0.672078
[5]	validation-auc:0.678571
[6]	validation-auc:0.672078
[7]	validation-auc:0.672078
[8]	validation-auc:0.678571
[9]	validation-auc:0.688312


[I 2022-06-02 17:38:57,297] Trial 65 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.9176078366773096, 'alpha': 3.811670981055984e-08, 'subsample': 0.32685372458015877, 'colsample_bytree': 0.5889008054893642}. Best is trial 2 with value: 1.0.
[I 2022-06-02 17:38:57,400] Trial 66 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:57,506] Trial 67 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.685065
[1]	validation-auc:0.668831
[2]	validation-auc:0.675325
[3]	validation-auc:0.678571
[4]	validation-auc:0.662338
[5]	validation-auc:0.659091
[6]	validation-auc:0.662338
[7]	validation-auc:0.662338
[8]	validation-auc:0.668831
[9]	validation-auc:0.668831


[I 2022-06-02 17:38:57,689] Trial 68 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.5492145188447105, 'alpha': 3.283533814438426e-08, 'subsample': 0.6153335693662416, 'colsample_bytree': 0.3682625129850939}. Best is trial 2 with value: 1.0.


[0]	validation-auc:0.685065
[1]	validation-auc:0.675325
[2]	validation-auc:0.675325
[3]	validation-auc:0.662338
[4]	validation-auc:0.659091
[5]	validation-auc:0.652597
[6]	validation-auc:0.662338
[7]	validation-auc:0.665584
[8]	validation-auc:0.665584
[9]	validation-auc:0.665584


[I 2022-06-02 17:38:57,864] Trial 69 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.6580316015396166, 'alpha': 6.883598829284126e-07, 'subsample': 0.6228478565708355, 'colsample_bytree': 0.2862307583339385}. Best is trial 2 with value: 1.0.
[I 2022-06-02 17:38:57,983] Trial 70 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.675325
[1]	validation-auc:0.672078
[2]	validation-auc:0.678571
[3]	validation-auc:0.665584
[4]	validation-auc:0.659091
[5]	validation-auc:0.668831
[6]	validation-auc:0.668831
[7]	validation-auc:0.675325
[8]	validation-auc:0.675325
[9]	validation-auc:0.678571


[I 2022-06-02 17:38:58,165] Trial 71 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.6860265224777591, 'alpha': 1.701416832868167e-06, 'subsample': 0.22795954794307194, 'colsample_bytree': 0.21989818994672242}. Best is trial 2 with value: 1.0.
[I 2022-06-02 17:38:58,263] Trial 72 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:58,363] Trial 73 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:58,463] Trial 74 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.681818
[1]	validation-auc:0.698052
[2]	validation-auc:0.681818
[3]	validation-auc:0.662338
[4]	validation-auc:0.650974
[5]	validation-auc:0.655844
[6]	validation-auc:0.646104
[7]	validation-auc:0.63961
[8]	validation-auc:0.646104
[9]	validation-auc:0.646104


[I 2022-06-02 17:38:58,657] Trial 75 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.2782343186829836, 'alpha': 1.1260843522320058e-07, 'subsample': 0.44153371026348237, 'colsample_bytree': 0.5695821312257756}. Best is trial 2 with value: 1.0.
[I 2022-06-02 17:38:58,752] Trial 76 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:58,870] Trial 77 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:38:58,966] Trial 78 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.691558
[1]	validation-auc:0.681818
[2]	validation-auc:0.675325
[3]	validation-auc:0.675325
[4]	validation-auc:0.668831
[5]	validation-auc:0.659091
[6]	validation-auc:0.665584
[7]	validation-auc:0.668831
[8]	validation-auc:0.668831
[9]	validation-auc:0.668831


[I 2022-06-02 17:38:59,148] Trial 79 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.606464415629099, 'alpha': 2.2720044884782113e-06, 'subsample': 0.6659196733590975, 'colsample_bytree': 0.2644511311786819}. Best is trial 2 with value: 1.0.
[I 2022-06-02 17:38:59,265] Trial 80 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.675325
[1]	validation-auc:0.665584
[2]	validation-auc:0.665584
[3]	validation-auc:0.675325
[4]	validation-auc:0.672078
[5]	validation-auc:0.681818
[6]	validation-auc:0.678571
[7]	validation-auc:0.678571
[8]	validation-auc:0.678571
[9]	validation-auc:0.688312


[I 2022-06-02 17:38:59,447] Trial 81 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.8951619903445687, 'alpha': 5.7330955073473684e-08, 'subsample': 0.5259431228588098, 'colsample_bytree': 0.23843509004212068}. Best is trial 2 with value: 1.0.


[0]	validation-auc:0.681818
[1]	validation-auc:0.672078
[2]	validation-auc:0.662338
[3]	validation-auc:0.659091
[4]	validation-auc:0.649351
[5]	validation-auc:0.646104
[6]	validation-auc:0.646104
[7]	validation-auc:0.646104
[8]	validation-auc:0.642857
[9]	validation-auc:0.642857


[I 2022-06-02 17:38:59,652] Trial 82 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.4545972696242773, 'alpha': 0.0003707695443857936, 'subsample': 0.2088909337818937, 'colsample_bytree': 0.6094877826472626}. Best is trial 2 with value: 1.0.
[I 2022-06-02 17:38:59,757] Trial 83 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.678571
[1]	validation-auc:0.672078
[2]	validation-auc:0.662338
[3]	validation-auc:0.662338
[4]	validation-auc:0.652597
[5]	validation-auc:0.646104
[6]	validation-auc:0.649351
[7]	validation-auc:0.646104
[8]	validation-auc:0.646104
[9]	validation-auc:0.649351


[I 2022-06-02 17:38:59,948] Trial 84 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.46237520752680994, 'alpha': 0.001022239329289067, 'subsample': 0.21740586659717553, 'colsample_bytree': 0.6133903254878398}. Best is trial 2 with value: 1.0.
[I 2022-06-02 17:39:00,043] Trial 85 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:39:00,149] Trial 86 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:39:00,247] Trial 87 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.685065
[1]	validation-auc:0.681818
[2]	validation-auc:0.685065
[3]	validation-auc:0.678571
[4]	validation-auc:0.678571
[5]	validation-auc:0.681818
[6]	validation-auc:0.675325
[7]	validation-auc:0.672078
[8]	validation-auc:0.672078
[9]	validation-auc:0.675325


[I 2022-06-02 17:39:00,421] Trial 88 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.8373727678018571, 'alpha': 8.607805681861056e-05, 'subsample': 0.3129248379752191, 'colsample_bytree': 0.5784009655497335}. Best is trial 2 with value: 1.0.
[I 2022-06-02 17:39:00,540] Trial 89 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:39:00,653] Trial 90 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:39:00,758] Trial 91 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:39:00,856] Trial 92 pruned. Trial was pruned at iteration 0.


[0]	validation-auc:0.685065
[1]	validation-auc:0.681818
[2]	validation-auc:0.675325
[3]	validation-auc:0.668831
[4]	validation-auc:0.668831
[5]	validation-auc:0.662338
[6]	validation-auc:0.665584
[7]	validation-auc:0.672078
[8]	validation-auc:0.665584
[9]	validation-auc:0.672078


[I 2022-06-02 17:39:01,042] Trial 93 finished with value: 0.8571428571428571 and parameters: {'booster': 'gblinear', 'lambda': 0.6891434997345821, 'alpha': 6.36119943223149e-07, 'subsample': 0.6174021642034098, 'colsample_bytree': 0.3846307632792385}. Best is trial 2 with value: 1.0.
[I 2022-06-02 17:39:01,144] Trial 94 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:39:01,244] Trial 95 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:39:01,340] Trial 96 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:39:01,438] Trial 97 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:39:01,556] Trial 98 pruned. Trial was pruned at iteration 0.
[I 2022-06-02 17:39:01,664] Trial 99 pruned. Trial was pruned at iteration 0.


Recall : %f 1.0
Best HyperParamteters : %f {'booster': 'dart', 'lambda': 8.669790267915281e-07, 'alpha': 1.1536382592082632e-08, 'subsample': 0.28469669085250404, 'colsample_bytree': 0.5566590888822147, 'max_depth': 7, 'min_child_weight': 10, 'eta': 0.003933096882701822, 'gamma': 0.0018015367008443472, 'grow_policy': 'lossguide', 'sample_type': 'weighted', 'normalize_type': 'forest', 'rate_drop': 8.829825957910605e-06, 'skip_drop': 5.189332052983798e-07}


### Recall
1.0

### Best HyperParamteters
'booster': 'dart', 

'lambda': 8.669790267915281e-07, 

'alpha': 1.1536382592082632e-08, 

'subsample': 0.28469669085250404, 

'colsample_bytree': 0.5566590888822147, 

'max_depth': 7, 

'min_child_weight': 10, 

'eta': 0.003933096882701822, 

'gamma': 0.0018015367008443472, 

'grow_policy': 'lossguide', 

'sample_type': 'weighted', 

'normalize_type': 'forest', 

'rate_drop': 8.829825957910605e-06, 

'skip_drop': 5.189332052983798e-07}